examples/shp/dot_product_sycl.cpp

```diff
@@ -0,0 +1,126 @@
+// #include <dr/shp.hpp>
+// #include <fmt/core.h>
+// #include <fmt/ranges.h>
+// #include <oneapi/dpl/algorithm>
+// #include <oneapi/dpl/async>
+// #include <oneapi/dpl/execution>
+// #include <oneapi/dpl/numeric>
+
+// #ifdef USE_MKL
+// #include <oneapi/mkl.hpp>
+// #endif
+
+#include <cstdio>
+#include <cstdlib>
+#include <iostream>
+#include <numeric>
+
+// #include <fmt/core.h>
+
+#include <CL/sycl.hpp>
+
+int main(int argc, char **argv) {
+
+  if (argc != 3) {
+    return 1;
+  }
+
+  std::size_t dev_num = std::atoi(argv[1]);
+  std::size_t vec_size = std::atoi(argv[2]);
+
+  if (vec_size == 0 || dev_num == 0) {
+    return 1;
+  }
+
+  double sycl_median = 0;
+
+  std::vector<sycl::device> devices;
+  for (auto const &this_platform : sycl::platform::get_platforms()) {
+    // you can also set the env variable
+    for (auto &dev : this_platform.get_devices(sycl::info::device_type::cpu)) {
+      devices.emplace_back(dev);
+    }
+  }
+
+  if (dev_num < devices.size()) {
+    std::cout << "There is not enough available devices" << std::endl;
+    std::cout << "Available devices= " << devices.size()
+              << ", requested devices= " << dev_num;
+    return 0;
+  }
+
+  vec_size = vec_size * 1000 * 1000ull;
+  using T = float;
+
+  std::vector<T> x(vec_size, 1);
+  std::vector<T> y(vec_size, 1);
+
+  std::vector<T> results(devices.size(), 1);
+
+  std::vector<sycl::queue> queues;
+  std::size_t remainder = vec_size % dev_num;
+  std::cout << devices.size() << std::endl;
+
+  std::vector<double> durations;
+  durations.reserve(dev_num);
+
+  for (std::size_t i = 0; i < devices.size(); i++) {
+    auto begin = std::chrono::high_resolution_clock::now();
+
+    queues.emplace_back(
+        sycl::queue(devices[i], sycl::property::queue::in_order{}));
+    std::size_t chunk_size = vec_size / dev_num;
```

Comment:
You get better (more uniform) distribution when last node is less loaded than the others:
chunk_size = vec_size / (dev_num - 1);
last_chunk_size = vec_size - chunk_size * (dev_num - 1);